### ISSUE TITLE
Are we going to support data from multiple analyses in ADC and/or iR+ repositories?

### ISSUE THREAD

**Issue by schristley** [ireceptor-plus/issues#43]
@bcorrie The AIRR Data Commons was primarily envisioned as an archive of the AIRR-seq data for a published study that can be queried and downloaded. However, with the ability of iR+ to run analyses on this data, we've opened a can of worms.

The basic iR+ workflow is a query takes data out of the database, puts it in a file, runs an analysis job on it, then visualizes the results. What happens if you want to run analysis on the output of an earlier analysis?

For example, query on rearrangements, run an analysis job to define clones on those rearrangements, now you want to run an analysis on that clone data like gene usage. Does that require loading the clone data into the repository?

According to the basic workflow, the answer is yes... Query the clones to take data out of the database, then run a gene usage analysis job on it.

If the data is not in the repository, how are we going to provide Stats API on it?

**Comment by bcorrie** [comment]
I see the Analysis API as quite simple in principle (although very powerful).

1. The user builds a query Q to select data set D. This query would be applied to the entire ADC - that is, it is potentially finding data from many places.
2. The user chooses an analysis tool T to run on this data.
3. The user uses the analysis API to run query Q to find data set D and run tool T on D - from a UI perspective, they essentially click `Go` - the analysis API finds the data using query Q, federates the data into a data set D, runs the analysis tool T, and gets some results R.
4. The user goes for coffee, and at some point in the future are notified that their analysis results R are done.
5. The user asks the analysis API to give them back the analysis results R and they do what they want with them...

**Comment by bcorrie** [comment]
> The basic iR+ workflow is a query takes data out of the database, puts it in a file, runs an analysis job on it, then visualizes the results. What happens if you want to run analysis on the output of an earlier analysis?

This is certainly desirable, but for me is outside the scope of the iR+ Analysis API.

**Comment by bcorrie** [comment]
> If the data is not in the repository, how are we going to provide Stats API on it?

I would argue we don't - if it isn't in a repository, the Stats API doesn't (and can't) do anything with it - since the Stats API is a repository API.

**Comment by bcorrie** [comment]
> For example, query on rearrangements, run an analysis job to define clones on those rearrangements, now you want to run an analysis on that clone data like gene usage. Does that require loading the clone data into the repository?

I would suggest that we would be well served by helping to facilitate the loading of objects that can be stored in an ADC or iR Plus repository based on the output of an iR Plus Analysis API job.

For example, I could see using the Analysis API to run a clone analysis tool on a set of repertoires in my own repository. The clone analysis tools creates me a set of AIRR Clone TSV or JSON files. I think we want to make it trivial to load the output of the Analysis API into an ADC repository.

This use case is me, as a data steward, using the Analysis API to help me curate data into my own repository. I could do the same for clone trees, etc.

**Comment by bcorrie** [comment]
> Furthermore, if the data is not in the repository, then providing such capability requires managing files, for both the input and output of jobs, across a series of jobs. This is essentially what VDJServer does... Are we reproducing VDJServer?

I think pipelines of jobs is outside of the scope of the Analysis API - at least in the simple model that I specify above. Not saying we can't or shouldn't tackle it, but that is not a deliverable from my perspective...

I could see VDJServer utilizing the iRPlus Analysis API to pull data from the ADC, perhaps doing some preprocessing to prepare for input into a more sophisticated pipeline, and then running a VDJServer analysis. But I don't think we want to reproduce VDJServer - we've already got one... https://www.youtube.com/watch?v=GYcopzJ-T9w 

I anticipate the iReceptor Gateway using the Analysis API to perform some sophisticated analyses on quite complex data sets federated from many repositories.

At the same time, I personally don't see the Analysis API pipelining many analyses together. I think there are tools that already fit that niche...

**Comment by schristley** [comment]
> I see the Analysis API as quite simple in principle (although very powerful).

Here's the challenge... There is the assumption that running tool T actually produces useful and complete analysis output, this assumption mostly incorrect, though it matters how you define "tool". Pretty much all immune repertoire analyses are workflows, requiring multiple steps and often multiple tools.

Take defining clones for example. It's not just running DefineClones in ChangeO; there are at least two additional steps that need to be performed. One, the rearrangements need to be filtered down to a valid set. Two, the Shazam tool needs to be run to determine a threshold parameter. And then at that point, there's a bunch of parameters that control the process, some which are specific to the organism (human, mouse).

**Comment by bcorrie** [comment]
I would define tool T in your above case as the three steps that you mention. The goal of tool T in this case is to generate clones using Change-O from data that is extracted from the ADC. This is different than just running Change-O. If you always have to do those three steps to run Change-O to define clones on rearrangement data from the ADC, then tool T performs those three steps. Internally to tool T a short workflow is implemented sure, but it is a fixed workflow that is run every time tool T is used. 

I would expect most tools we would support as needing to do some sort of preprocessing that will need to be performed to get the data from the format that it comes from the ADC into the form that is actually required by the tool that is being run. 

WRT parameters, I think that is unavoidable - there will always be a need to pass some sort of parameters to the application - but Tapis already has this capability, so I don't think that is a real issue is it?

**Comment by schristley** [comment]
> I would define tool T in your above case as the three steps that you mention.

Ok, well maybe we just have a terminology difference. I think of tools as Change-O is a tool, sonar is a tool, allakazam is a tool, etc. Some tools maybe for a specific analysis while some tools may support many different analyses, and some analyses require using multiple tools. The way you describe it makes it sound like you want an analysis, so I would rephrase your 2) as this:

2. The user chooses an analysis A to run on this data.

That has different implications because analysis A can be provided by different (my definition) tools. This is closer to my initial thoughts about the analysis API, in that it defined analyses but was essentially agnostic to the actual tools being used. Kind of how the stats API is right now, it doesn't say "use tool X to calculate gene usage", it just defines an API to get gene usage for a repertoire and let's the backend service free to choose how. My thought is the analysis API for gene usage would be almost identical to the stats API.

@bcorrie Does this still correspond with your thinking?

**Comment by schristley** [comment]
> I would suggest that we would be well served by helping to facilitate the loading of objects that can be stored in an ADC or iR Plus repository based on the output of an iR Plus Analysis API job.

I'm not sure that I'm comfortable with loading up all analysis output into the repository. While it's possible for some analyses like defining clones, other analysis output doesn't fit into any of the AIRR objects.

Even in the cases where it's possible, e.g. defining clones, if a user runs 100 analyses with different parameters, do we have to load up all them?

**Comment by bcorrie** [comment]
> That has different implications because analysis A can be provided by different (my definition) tools. This is closer to my initial thoughts about the analysis API, in that it defined analyses but was essentially agnostic to the actual tools being used. Kind of how the stats API is right now, it doesn't say "use tool X to calculate gene usage", it just defines an API to get gene usage for a repertoire and let's the backend service free to choose how. My thought is the analysis API for gene usage would be almost identical to the stats API.
> 
> @bcorrie Does this still correspond with your thinking?

I admit I was thinking along the line of "Tools". For something "simple" like gene usage counts, if one defines what is meant by a count, any tool can do the count. But it doesn't seem like that extends very well to Tools and Analysis. It seems to me like the researcher is going to care very much about which Tool does the Analysis???

I don't think that necessarily precludes doing what you say - that is, we as iR+ could choose to support a set of specific Tool that perform a specific Analysis with a 1-1 relation between Tool and Analysis for iR+ implementation purposes. That keeps the deliverable for iR+ more tractable.

I think it would be nice if the back end that implements that has more flexibility (e.g. Tapis and its App/Sys/Platform/File object structure) so we have much more power behind the scenes - but we can keep it simple from an external iR+ API perspective if we want. That is don't make the Tapis part an iR+ deliverable - it is an implementation detail behind the scenes...

**Comment by bcorrie** [comment]
> > I would suggest that we would be well served by helping to facilitate the loading of objects that can be stored in an ADC or iR Plus repository based on the output of an iR Plus Analysis API job.
> 
> I'm not sure that I'm comfortable with loading up all analysis output into the repository. While it's possible for some analyses like defining clones, other analysis output doesn't fit into any of the AIRR objects.
> 
> Even in the cases where it's possible, e.g. defining clones, if a user runs 100 analyses with different parameters, do we have to load up all them?

I agree 100% - we don't want to do that in the general case... 

I was thinking more along the lines of if I run the iR+ Clone Tool/Analysis, the output of that tool (an AIRR Clone file) would be easy to load into an iReceptor Turnkey repository. This wouldn't necessarily be an automated process in any way, or even a general requirement.

I just think that if I was a data curator, and I was running an iReceptor Turnkey, we (the iR+ team) would want to make it easy for that curator to do the following:

1. Go to the Gateway, choose a Repertoire in my repository
2. Run the iR+ Clone analysis on the Rearrangement data in that Repertoire to generate an AIRR Clone file
3. Download that AIRR Clone file to my Turnkey
4. Load the clone file for that repertoire into my repository - viola, now I have clones for my repertoire, using the Gateway and the analysis API to generate them

Just making sure the dots are connected - but definitely not wanting that as a general requirement...

**Comment by schristley** [comment]
> I just think that if I was a data curator, and I was running an iReceptor Turnkey, we (the iR+ team) would want to make it easy for that curator to do the following:

I agree. I want the same thing from a curator perspective for VDJServer too.

**Comment by schristley** [comment]
> It seems to me like the researcher is going to care very much about which Tool does the Analysis???

Sure. I might say it more precisely that researchers care about the Method used, that generally corresponds to different Tools, but some simpler analyses essentially have a single method. So for mutation analysis, which is essentially comparing a sequence to its germline, it's pretty much the same method, likewise for gene usage or diversity. On the other hand, lineage tree reconstruction is an approximate calculation, so multiple methods are available.

**Comment by schristley** [comment]
> > Even in the cases where it's possible, e.g. defining clones, if a user runs 100 analyses with different parameters, do we have to load up all them?
> 
> I agree 100% - we don't want to do that in the general case...

Just to clarify my point, in this specific example, if the clones are not loaded into the repository, then how do we display stats like gene segment usage and so on? Or run another downstream analysis like mutations or lineage reconstruction?

**Comment by bcorrie** [comment]
Since the Stats API is a repository API, it can only do Stats on things in the repository. So if there aren't clones in the repository for a specific repertoire, then the Stats API doesn't help... If a repository stores an entity, then we can define a Stats API entry point for that entity. We have done this for rearrangements, we are working on the clones, I assume we will have a cell Stats, and possibly, assuming we get to the point of storing lineage trees in the repository, then we could do Stats on lineage trees. All of these objects are in the AIRR spec, so it makes sense that they would/could be stored in a repository, and therefore it makes sense there would/could be a Stats API for each object type.

**Comment by bcorrie** [comment]
If a repertoire only has rearrangements and does not have clones in the repository, then the Stats API doesn't help, but the analysis API could/should...

This depends on what iR+ defines as to what an `Analysis` does. At an abstract level an `Analysis` takes a set of entities in an ADC repository and performs some sort of analysis on them, producing some sort of output. That is the minimal viable product for an `Analysis` I would think. From a functional perspective, we need to think of what functions we want an `Analysis` to perform, what its inputs would be, and what its outputs would be. 

If we map the functional equivalent of an `Analysis` to the functionality of a single tool (e.g Presto, Change-O, tcr_dist, sum_rep, ...), then you have lots of `Analysis` entities that you have to define. This is possible, but as you suggest, you need multiple tools to be run in a pipeline to functionally accomplish what you would think of as something like a "complete clone analysis". So if iR+ wants to be able to do a "complete clone analysis" and iR+ doesn't want to have to implement the management of analysis pipelines, I don't think we want to think of an `Analysis` at the tool level.

What I would suggest as a starting point would be to come up with a small number of high-value functional analyses, focus on those, and come up with a couple of prototypes. We were getting pretty close to that at our last hackathon at the tool level, maybe it is time to do some definition work for a couple of `Analysis` tools and then have that second hackathon we were talking about 8-)

Thinking along the lines of what I believe you were suggesting for a clone analysis, we should try and determine what we think of as a functional `Clone Analysis`. This is the "http://analysis.ireceptorplus.org/analysis/v1/clone" entry point. As you say this is more than running just Change-O.  I would start with what do we want the output of that functional analysis to be? The output could be a simple clone file and nothing more, but even to get to that you have to run multiple tools. And "functionally" that is probably incomplete. You probably want some summary stats like diversity metrics, gene usage for the clones in that repertoire, etc. The output might be a clone file, some stats in a tab delimited file, and maybe some pretty graphs of gene usage. So "functionally" the `Clone Analysis` is likely going to consist of running a bunch of existing tools to produce a desired output. A mini-pipeline, true, but a fixed pipeline that is run over and over again...

**Comment by bcorrie** [comment]
> > > Even in the cases where it's possible, e.g. defining clones, if a user runs 100 analyses with different parameters, do we have to load up all them?
> > 
> > 
> > I agree 100% - we don't want to do that in the general case...
> 
> Just to clarify my point, in this specific example, if the clones are not loaded into the repository, then how do we display stats like gene segment usage and so on? Or run another downstream analysis like mutations or lineage reconstruction?

I think whenever you use a term like downstream analysis, you are talking about pipelines. That is VDJServer's world for sure, but I am not sure we want the iR+ Analysis API to handle that... 

For me I see the iR+ Analysis API computing on data extracted from repositories and generating analysis output (but not really doing anything with that output). It is the clients job to do something with the output. For example:
- a `Clone Analysis` which computes on rearrangements (extracted from a repository) and generates clones
- a `Lineage Tree` analysis that computes on clones (extracted from a repository) and generates clonal lineage trees
- a `Mutation Analysis` that computes on clones (extracted from a repository) and generates a mutation analysis report
- a `Lineage Tree Stats` analysis that computes on clonal lineage trees (extracted from a repository) and generates nice Stats about lineage trees.
- maybe even a `Clone Tree Analysis` that computes on rearrangements (extracted from a repository) and generates clonal lineage trees, generating the clones internally as a process required to generate the clonal lineage trees. Basically a monolithic `Analysis` that internally is a mini-pipeline... 

If the iReceptor Gateway made use of `Clone Analysis` on the rearrangements from a specific repertoire using the iR+ Analysis API, then it would need to determine what to do with the outputs of the `Analysis`. It might just let the user download the output. Or maybe it figures out how to present the output to the user in a nice way.

I could also see VDJServer using the iR+ Analysis API to compose a set of individual `Analyses` into an analysis pipeline, with VDJServer managing the input and output of the different steps of the analysis in that pipeline. I think this is already what VDJServer does, correct? For example, assuming there was a `Presto` and `igblast` iR+ analysis, VDJServer could use the iR+ Analysis API to run `Presto` -> `igblast` -> `Clone Analysis` -> `Lineage Tree`, including a branch on the output of `Clone Analysis` -> `Mutation` managing the input/output of each step as appropriate.

The iR+ Analysis API simplifies everyone's life by providing a small number of highly valuable `Analysis` capabilities that compute on data extracted from ADC repositories. 

What I would be concerned about is having the iR+ Analysis API be responsible for the pipeline part, in particular the plumbing of joining multiple `Analyses` together into a complex pipeline.

**Comment by schristley** [comment]
> This depends on what iR+ defines as to what an `Analysis` does.

Excellent, yes I agree with the thoughts you've put down. When it comes to functionality, we should be able to do everything you say, define high-value analyses, define the outputs, and etc.

What would be useful to think about though is maintainability and flexibility, this is something that I still struggle/debate with the best approach. My terminology is specific to Tapis but I think you will get the idea. So think about a Tapis app, like we were doing with the hackathon. Should a Tapis app be a "pipeline" that does a complete analysis, or should it be a strict wrapping of a tool?

In the former case, the Tapis app does everything needed to produce the analysis output, including running multiple tools and etc. In the latter case, the Tapis app essentially acts like running the tool from the command line, and thus is usually insufficient to do a complete analysis, multiple Tapis apps might need to be run in succession to get the final analysis output, passing data from one to the next.

Now you might think that the former, a Tapis app as a pipeline is the "way to go". This is essentially what VDJServer has now, but it has problems:

1. It's inflexible. You tend to be stuck with the analysis you initially code.
2. You cannot really update tools. You have to update whole pipelines.
3. They are complex to code and complex to scale up and/or parallelize.

Regarding 1), you might say, well, just expose parameters to make the pipeline flexible. But what you find is that it's not just a different value being passed to a tool, you end up writing branching logic and soon your "single" pipeline turns into 5 or 6 pipelines to handle the flexibility. Now imagine trying to update (and test) that complicated "pipeline" with updated versions of tools. This is a bit of the complexity and maintenance problem that I have with VDJServer right now...

In some sense, with the Tapis app as a pipeline, you don't need an analysis API. Running a job with that app essentially becomes the "analysis/v1/clone" end point. True, the API does provide an abstraction away from Tapis or whatever backend technology but the API is essentially a pass-through. The parameters on the API are the same parameters on the Tapis app.

What's so great about having the Tapis app be just a wrap of a tool.

1. Very flexible. The client can run the tool in any possible way it wants.
2. Direct mapping between Tapis app parameters and tool parameters.
3. Making a new tool version is straightforward, update the Tapis app and publish it as a new version. The older versions of the tools can still be available.
4. It should be easier to scale up and/or parallelize.

But what is the downside, which is major, is that we would need to "compose" pipelines to produce the complete analyses that we are interested in providing. That is, like today where you write a BASH script that runs the command line tools in succession, we would need a similar capability.

**Comment by schristley** [comment]
> For me I see the iR+ Analysis API computing on data extracted from repositories and generating analysis output (but not really doing anything with that output). It is the clients job to do something with the output.

Ok, I'm glad you verbalized that because that was one of the things I was struggling with for iR+.

> I could also see VDJServer using the iR+ Analysis API to compose a set of individual `Analyses` into an analysis pipeline, with VDJServer managing the input and output of the different steps of the analysis in that pipeline. I think this is already what VDJServer does, correct?

Actually no. While VDJServer does have 3 main steps: pre-processing (presto/vdjpipe), gene annotation (IgBlast), analysis (RepCalc), they are essentially monolithic pipelines and not very flexible. Particularly analysis which is this big complex app. It sounded like a good idea at the time, generate all the analysis in one run! But it's turned out to be a maintenance headache and inflexible. I actually want to convert VDJServer to a more flexible system.

**Comment by schristley** [comment]
> The iR+ Analysis API simplifies everyone's life by providing a small number of highly valuable `Analysis` capabilities that compute on data extracted from ADC repositories.

Yes, exactly, that makes the GUI integration be like VDJServer today, you pretty much push a button and you get this high value analysis. But it does re-introduce the inflexibility issue, in order to keep it simple there may only be a few parameters to adjust how that analysis is done. They essentially become the set of "standard iR+ curated analyses".

**Comment by bcorrie** [comment]
Off the top of my head, I think I would tend to lean towards starting with a Tapis App as a Tool wrapper as a base implementation layer. In particular for dockerized tools for which we can use the singularity approach that we were experimenting with in the hack-a-thon. Composition of tools into more complex `Analyses`, even internally could be deferred a bit. 

I am basing this on the assumption that there might be enough high value Tools that we could use directly without having to compose tools into more complex pipelines? Or if nothing else, the composition is simple enough that it can be done within the Tapis wrapper with a bit of bash magic??? 8-)

Even the Change-O use case that you mentioned above I think could be thought of as wrapping "immcantation" as a tool in the sense that you would Tapis wrap all of immcantation's docker container, then write a bit of internal shell magic to filter, run Shaazam, extract the parameters/threshold that Shaazan suggests, run Change-O, voila? 8-) I know it is not that easy, but I think it might be worth trying?

**Comment by felixbreden** [comment]
hi brian and scott
very interesting discussion - would it make sense to summarize this at the next WP4 meeting?  assuming most iR+ participants won;t read it all and think about it.  in all of this can we prioritize 2 analysis "tools" and get these implemented and we can see what it means to do all this work and how valuable users find it?  While reading about the standard use cases/tools, and the question of whether the output gets stored in the ADC, it sounds like it is agreed that clonal analysis gets stored in ADC, or is that just if the repository does it on a new repository, and not if a user does it.  hmmm.  what I am leading up to, if we have a standard clonal analysis that will aggregate repertoires into clones, could we also have a standard "germline analysis" that too each and every repertoire run it through VDJbase and store the output in ADC?  just dreaming I guess - looking forward to the next steps!

**Comment by schristley** [comment]
> I am basing this on the assumption that there might be enough high value Tools that we could use directly without having to compose tools into more complex pipelines? Or if nothing else, the composition is simple enough that it can be done within the Tapis wrapper with a bit of bash magic??? 8-)

Yes, I think that assumption is valid for a number of cases. For example, the [gene usage script](https://github.com/ireceptor-plus/tapis/blob/alakazam/apps/alakazam/4.1/common/gene_usage.R) that I wrote for Alakazam falls into that category. In fact, you MUST write a script because Alakazam is an R package and does not have a command line interface. Also if you look at that code more closely, it doesn't parameterize what usage counts to generate, it just calculates them all, which is "simpler" than the stats API.

This brings up an interesting side point. Even if we talk about a Tapis App wrapping a Tool, that's pretty imprecise. Alakazam is actually like 30+ functions for roughly 5 main purposes. Change-o is actually 8 separate python programs. That opens the question of granularity, do we have a single Alakazam Tapis app that does all of the functionality with a whole bunch of parameters? On the other end, do we have 8 Tapis apps for each Change-o program? I guess where I am going is that making any individual Tapis App too "big" turns into the complex, inflexible problem. Too "simple" and we need to have pipelines/workflows to stitch them together. We need to find the goldilocks region that is just right.

**Comment by schristley** [comment]
> very interesting discussion - would it make sense to summarize this at the next WP4 meeting?

yes, no maybe ;-D As the majority of the consortium doesn't do immune repertoire analysis, I think they just want a decision to be made. For the ones that do, it's unclear if they will really use the analysis API. I imagine they don't know themselves until there is something concrete for them to use and can provide feedback on.

Brian and I have only been talking about technical thingies, the "political" issues are still unresolved. Those have a big influence on what we will actually do.

> can we prioritize 2 analysis "tools" and get these implemented and we can see what it means to do all this work and how valuable users find it?

We have 1 now with what I've done with Alakazam, but these are "simple" analyses. Defining clones is really the first "complex" analysis, and actually I created [this](https://github.com/ireceptor-plus/tapis/tree/alakazam/apps/changeo/0.4) almost a year ago but there was too much WP1/WP2 work, the bandwidth wasn't there. Now that clones are "hot" again, we should be able to move quickly.

> could we also have a standard "germline analysis" that too each and every repertoire run it through VDJbase and store the output in ADC?

Yes, I think we should "wrap" the VDJbase workflow so this can be done. I'm not familiar with the workflow to know if it's "complex" and should be broken down into steps, or if there are "checkpoints" where the user would want to do some validation of the results before moving on to the next step.

**Comment by schristley** [comment]
We've gone with the design of providing a set of Tapis analysis apps, and clients directly suppose analysis jobs through Tapis. Client can provide support for workflows/pipelines if they want (e.g. VDJServer), but the iR+ platform won't have an explicit workflow capability.
